src/crypto-providers/ledgerCryptoProvider.ts

```diff
@@ -455,6 +458,38 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
     }) : undefined
   )
 
+  const prepareScriptDataHash = (
+    scriptDataHash: Buffer | undefined,
+  ): string | undefined => scriptDataHash?.toString('hex')
+
+  const prepareCollateralInput = (
+    collateralInput: TxTypes.Collateral, path: BIP32Path | null,
+  ): LedgerTypes.TxInput => {
+    if (collateralInput.index < 0 || collateralInput.index > Number.MAX_SAFE_INTEGER) {
+      throw Error(Errors.InvalidCollateralInputError)
+    }
+    return {
+      path,
+      txHashHex: collateralInput.transactionId.toString('hex'),
+      outputIndex: Number(collateralInput.index),
+    }
+  }
+
+  const prepareRequiredSigner = (
+    requiredSigner: TxTypes.RequiredSigner, signingFiles: HwSigningData[],
+  ): LedgerTypes.RequiredSigner => {
+    const path = findSigningPathForKeyHash(requiredSigner, signingFiles)
+    return path
+      ? {
+        type: LedgerTypes.TxRequiredSignerType.PATH,
+        path,
+      }
+      : {
+        type: LedgerTypes.TxRequiredSignerType.HASH,
+        hash: requiredSigner.toString('hex'),
+      }
+  }
+
   const prepareAdditionalWitnessRequests = (
     mintSigningFiles: HwSigningData[],
     multisigSigningFiles: HwSigningData[],
```

Comment:
How should we modify `prepareAdditionalWitnessRequests` for Alonzo? In Alonzo txs, successful Plutus script validation may require eg. three 1852 signatures, but if the tx contains only one input and one collateral input, there is no way how to force hw-cli to also create the third 1852 signature (except for calling `witness` call manually). Ideally, the signers required by Plutus script should be listed in `requiredSigners`, but I'm afraid we cannot rely on this always being the case.

The best solution I can think of right now is to modify it like this:
```ts
let files = [...mintSigningFiles ...multisigSigningFiles]
if (signingMode === SigningMode.PLUTUS_TRANSACTION) {
    files = [...files, ...paymentSigningFiles, ...stakeSigningFiles] 
}
return files.map((f) => f.path)
```

This way, some unwanted 1852 signatures will be created if the user provides superfluous signing files, but this is limited only to the Plutus signing mode (where the user sees all witness requests on the hw wallet screen). Furthermore, we are already doing this for 1854 signing files in multisig signing mode.

@janmazak what do you think?

Reply:
If superfluous witnesses appear only if the user explicitly supplies some signing files, I see no real problem, it's a bit of an inconvenience at most. It might be resolved in the future if it bothers someone.

Perhaps add a comment explaining the situation.

Reply:
Dex guys notified us that Plutus scripts can not access the signatures - they see only required signers (and the other tx items). Thanks to this, I believe we don't have to modify `prepareAdditionalWitnessRequests`. However, I will add a comment explaining this, in order not to confuse someone in the future.